enterprise/tpa_pipeline.py

```diff
@@ -0,0 +1,153 @@
+
+    Args:
+        request: HttpRequest object containing request data.
+        enforcement_location (str): the point where to see data sharing consent state.
+        argument can either be "optional", 'at_login' or 'at_enrollment'
+    """
+    running_pipeline = request.session.get('partial_pipeline')
+    if running_pipeline:
+        customer = get_enterprise_customer_for_request(request)
+        return customer and customer.enforces_data_sharing_consent(enforcement_location)
+    return False
+
+
+def active_provider_requests_data_sharing(request):
+    running_pipeline = request.session.get('partial_pipeline')
+    if running_pipeline:
+        customer = get_enterprise_customer_for_request(request)
+        return customer and customer.requests_data_sharing_consent
+    return False
+
+
+def get_consent_status_for_request(request):
+    customer = get_enterprise_customer_for_request(request)
+    user = request.user
+    try:
+        return UserDataSharingConsentAudit.objects.get(
+            user=user,
+            enterprise_customer=customer,
+        )
+    except UserDataSharingConsentAudit.DoesNotExist:
+        return None
+
+
+@partial.partial
+def verify_data_sharing_consent(backend, user, **kwargs):
+    """
+    Checks to ensure that the user has provided data sharing consent
+    if the active SSO provider requires it; if not, then the user will
+    be redirected to a page from which they can provide consent.
+    """
+
+    def redirect_to_consent():
+        """
+        Method redirects the user to a page from which they can provide
+        required data sharing consent before proceeding in the pipeline
+        """
+        return redirect(reverse('grant_data_sharing_permissions'))
+
+    customer = get_enterprise_customer_for_running_pipeline({'backend': backend.name, 'kwargs': kwargs})
+    if customer is None:
+        return
+
+    if not customer.requests_data_sharing_consent:
+        return
+
+    try:
+        consent = UserDataSharingConsentAudit.objects.get(user=user, customer=customer)
+    except UserDataSharingConsentAudit.DoesNotExist:
+        return redirect_to_consent()
+
+    if (not consent.enabled) and customer.enforces_data_sharing_consent(EnterpriseCustomer.AT_LOGIN):
+        return redirect_to_consent()
+
+
+def set_data_sharing_consent_record(backend, user, data_sharing_consent=None, **kwargs):
+    """
+    If the pipeline produced a command to explicitly set the data sharing consent
+    record a particular way, set it that way.
+    """
+    if data_sharing_consent is None:
+        return
+
+    customer = get_enterprise_customer_for_running_pipeline({'backend': backend.name, 'kwargs': kwargs})
+    if customer is None:
+        return
+
+    consent, _ = UserDataSharingConsentAudit.objects.get_or_create(user=user, enterprise_customer=customer)
+
+    if data_sharing_consent:
+        consent.enable()
+    else:
+        consent.disable()
```

Comment:
I think that if you change `UserDataSharingConsentAudit` to use `HistoricalRecords`, you can get rid of this need to call enable()/disable()/save() methods, and just use the much nicer single-statement

```python
UserDataSharingConsentAudit.objects.update_or_create(
    user=user,
    enterprise_customer=customer,
    defaults={'state': ENABLED if data_sharing_consent else DISABLED},
)
```

Same thing below in `GrantDataSharingPermissions`.`post()`